source/transfer_list.rst

```diff
@@ -47,7 +47,10 @@ Transfer list header
 A TL must begin with a TL header. The layout of the TL header is shown in
 :numref:`tab_tl_header`.  The presence of a TL header can be verified by
 inspecting the signature field which must contain the 0x6e_d0ff value.  The
-version field determines the contents of the handoff start header.
+version field determines the contents of the handoff start header. The version
+will only be changed by an update to this specification when new TL header or
+TE header fields are defined (i.e. not when allocating new tag IDs), and all
+changes will be backwards-compatible to older readers.
```

Comment:
As we discussed, that is not true if we decide to change the TE header size, since we would not want to be stuck with what would look like a strange layout in a 16-byte header. Can we drop this from here and deal with it once we have some adoption, as agreed?

So just drop ", and all changes will be backwards-compatible to older readers."

We can add it in 6 months, say, once we are confident.

Reply:
This text is specific to the `version` field, and I think the point of that field is specifically to manage backwards-compatible changes (which we might want to make at some point in the future, not necessarily in 6 months). This is not meant to contradict what we discussed, but if we decide to make a backwards-incompatible change in 6 months, then this version field becomes meaningless (e.g. we'll basically start at 1 again for the new definition of the spec).

For backwards-incompatible changes, if you want a way to programmatically distinguish those, I'd suggest altering the signature field instead.

Reply:
That doesn't sound right to me. How is the text specific to the version field?

"and all changes will be backwards-compatible to older readers" seems to refer to the entire spec, not just the version field.

Changing back to a 16-byte TE header does not require starting the spec from scratch. That sort of comment worries me. The TE header is a fairly minor change in the text, but it would be an incompatible change.

We specifically discussed that if we want to change the TE header that it would *not* be backwards-compatible, but we would simply change it. I am fine with incrementing the version in that case (so that version 1 just disappears and no one uses it). But writing a new spec is absolutely not what we discussed.

Reply:
> How is the text specific to the version field?

...because it talks about the version field? I mean, the part you quoted literally starts with "version field determines ...". ;)

Sorry, I don't feel like we actually have a disagreement in what the plan is here, there's just a misunderstanding in how that would be written down. I don't think the spec needs to make provisions in the text itself on how backwards-incompatible changes happen -- because if you're making a backwards-incompatible change, then you can by definition ignore anything that has been written before anyway. So while we all agreed on the 6 months thing out-of-band and I'm not trying to renege on that in any way, I don't think there's a point in writing any of that down in the text itself (and I'm worried that doing so might hurt adoption).

I do however want to make sure that the rules for backwards-compatible changes are well-defined, because that has important consequences to people who implement the spec right now. So this piece of text is supposed to clarify those rules (which IIRC we had discussed and agreed on in some other comment thread earlier in this PR). And the version field can only be used for either backwards-compatible or backwards-incompatible changes, not both, so that's why this needs to define that the changes tracked by that mechanism must be backwards-compatible. If we decide that we want to make a backwards-incompatible change, like I said we can just change the signature and then older readers will treat it as a completely unknown thing, which is the behavior you want for backwards-incompatible changes. (This doesn't mean that the change needs to be big in any way, copying the whole spec and just changing the header size is still a totally valid backwards-incompatible change, but older readers still need to treat that like a totally different and unknown thing.)

Does that make sense?

Reply:
I'm not really keen on changing the signature, since then the old one lives on. It needs to vanish and we don't want the old one to continue at all. But I don't think it's the end of the world, if you are going to insist on it living on in some form. Would we have to mention the old version at all?

Also we don't need to 'copy' the spec. The change would just be a PR.

So does this sound right?

After some projects have adopted this, if something comes up so that we want to increase the TE header size (e.g. to 16 bytes), we will do so in an incompatible manner, with a PR, with communication to all projects. This is to avoid needing to support both implementations. Any such change should be a fairly minor code change for each project, just adjusting a structure and the usage of align8().

We don't need this in the spec, but I do want to have it clearly agreed. I can do a PR for the readme, perhaps

Reply:
Well, then don't change the signature, that's fine by me too. I'm just saying it should not change the version field, because I think the version field should only track backwards-compatible changes.

> After some projects have adopted this, if something comes up so that we want to increase the TE header size (e.g. to 16 bytes), we will do so in an incompatible manner, with a PR, with communication to all projects. This is to avoid needing to support both implementations. Any such change should be a fairly minor code change for each project, just adjusting a structure and the usage of align8().

How about putting it this way: Within the first X months [6 months? 1 year? I don't care that much about the exact number but it should be clearly bounded, so that eventually adopters can have certainty that this is a stable format] after official release of this spec, if we find any serious flaw in the design that we feel needs to be addressed to help this format succeed and be universally useful [this could be adding extra fields to the TE header, or anything else], we reserve the option to make a backwards-incompatible change to the spec [this could be a big change or a small change depending on the nature and severity of the problem and we'll discuss the details and pros/cons of each option in the context of the specific problem when it comes up]. We will update the spec with that, coordinate with all projects already using it to switch to the new version, and then forget the older version ever existed. After that time period, we will consider the spec stable and do whatever possible to address any future problems or new requirements in a backwards-compatible manner according to the backwards-compatibility / version field rules defined in the spec.

Reply:
Yes that is good with me. I'll do a readme update tomorrow then.